src/realm/replication.hpp

```diff
@@ -474,18 +474,25 @@ inline void Replication::select_table(const Table* table)
 
 inline bool Replication::select_collection(const CollectionBase& coll)
 {
+    ConstTableRef table = coll.get_table();
+    ColKey col_key = coll.get_col_key();
+    ObjKey obj_key = coll.get_owner_key();
+    auto path = coll.get_stable_path();
+
+    bool newly_created_object = check_for_newly_created_object(coll.get_owner_key(), table.unchecked_ptr());
     if (CollectionId(coll) != m_selected_collection) {
-        do_select_collection(coll); // Throws
+        do_select_collection(table, col_key, obj_key, std::move(path)); // Throws
     }
-    return !m_newly_created_object;
+    return !newly_created_object;
 }
 
-inline bool Replication::select_obj(ObjKey key)
+inline bool Replication::select_obj(ObjKey key, const Table* table)
 {
-    if (key != m_selected_obj) {
-        do_select_obj(key);
+    bool newly_created = check_for_newly_created_object(key, table);
```

Comment:
There doesn't appear to be any obvious reason we have to check this every time rather than caching it and only checking when selecting an object?

Reply:
That is a valid optimization, changed.